src/lib/optimizer/strategy/join_ordering_rule.cpp

```diff
@@ -0,0 +1,68 @@
+#include "join_ordering_rule.hpp"
+
+#include "expression/expression_utils.hpp"
+#include "logical_query_plan/projection_node.hpp"
+#include "optimizer/dp_ccp.hpp"
+#include "optimizer/join_graph.hpp"
+#include "utils/assert.hpp"
+
+namespace opossum {
+
+JoinOrderingRule::JoinOrderingRule(const std::shared_ptr<AbstractCostEstimator>& cost_estimator)
+    : _cost_estimator(cost_estimator) {}
+
+std::string JoinOrderingRule::name() const { return "JoinOrderingRule"; }
+
+bool JoinOrderingRule::apply_to(const std::shared_ptr<AbstractLQPNode>& root) const {
+  /**
+   * Dispatch _perform_join_ordering_recursively() and fix the column order afterwards
+   */
+
+  Assert(root->type == LQPNodeType::Root, "JoinOrderingRule needs root to hold onto");
+
+  const auto expected_column_order = root->column_expressions();
+
+  auto result_lqp = _perform_join_ordering_recursively(root->left_input());
+
+  // Join ordering might change the output column order, let's fix that
+  if (!expressions_equal(expected_column_order, result_lqp->column_expressions())) {
+    result_lqp = ProjectionNode::make(expected_column_order, result_lqp);
+  }
+
+  root->set_left_input(result_lqp);
+
+  // Figuring out whether the JoinOrderingRule changed the LQP is hard, and the rule should be applied only once,
+  // anyway. So, return false.
```

Comment:
Isn't that made sure by the execution policy? Returning a false result just because you (currently) know what the caller does with it smells.

Reply:
I do it because "Figuring out whether the JoinOrderingRule changed the LQP is hard"

I put the "(and the rule should be applied only once, anyway)" bit in parentheses.